### PR TITLE
feat!: use nullable Int64 and boolean dtypes if available

### LINKED ISSUE #426: support google-cloud-bigquery 3.x (test with pre-release dependencies)
There is currently a beta 1 pre-release for google-cloud-bigquery 3.x, it'd be great if we could expand our support for it already -- or at least run some tests to see what breaks.

It does change some default dtypes, so there is a good chance at least some tests will fail.

### LINKED ISSUE #429: tests.system.test_gbq.TestReadGBQIntegration: test_should_properly_handle_valid_integers failed
This test failed!

To configure my behavior, see [the Flaky Bot documentation](https://github.com/googleapis/repo-automation-bots/tree/main/packages/flakybot).

If I'm commenting on this issue too often, add the `flakybot: quiet` label and
I will stop commenting.

---

commit: 928e47b4bc1a8d933f48206ba98e3d22726876c8
buildURL: [Build Status](https://source.cloud.google.com/results/invocations/bbab8f02-1708-462f-8c90-4285e1304f0a), [Sponge](http://sponge2/bbab8f02-1708-462f-8c90-4285e1304f0a)
status: failed
<details><summary>Test output</summary><br><pre>self = <system.test_gbq.TestReadGBQIntegration object at 0x7f3e6843b040>
project_id = 'precise-truck-742'

    def test_should_properly_handle_valid_integers(self, project_id):
        query = "SELECT CAST(3 AS INT64) AS valid_integer"
        df = gbq.read_gbq(
            query,
            project_id=project_id,
            credentials=self.credentials,
            dialect="standard",
        )
>       tm.assert_frame_equal(df, DataFrame({"valid_integer": [3]}))
E       AssertionError: Attributes of DataFrame.iloc[:, 0] (column name="valid_integer") are different
E       
E       Attribute "dtype" are different
E       [left]:  Int64
E       [right]: int64

tests/system/test_gbq.py:171: AssertionError</pre></details>

### LINKED ISSUE #430: tests.system.test_gbq.TestReadGBQIntegration: test_should_properly_handle_valid_longs failed
This test failed!

To configure my behavior, see [the Flaky Bot documentation](https://github.com/googleapis/repo-automation-bots/tree/main/packages/flakybot).

If I'm commenting on this issue too often, add the `flakybot: quiet` label and
I will stop commenting.

---

commit: 928e47b4bc1a8d933f48206ba98e3d22726876c8
buildURL: [Build Status](https://source.cloud.google.com/results/invocations/bbab8f02-1708-462f-8c90-4285e1304f0a), [Sponge](http://sponge2/bbab8f02-1708-462f-8c90-4285e1304f0a)
status: failed
<details><summary>Test output</summary><br><pre>self = <system.test_gbq.TestReadGBQIntegration object at 0x7f3e627c6550>
project_id = 'precise-truck-742'

    def test_should_properly_handle_valid_longs(self, project_id):
        query = "SELECT 1 << 62 AS valid_long"
        df = gbq.read_gbq(
            query,
            project_id=project_id,
            credentials=self.credentials,
            dialect="standard",
        )
>       tm.assert_frame_equal(df, DataFrame({"valid_long": [1 << 62]}))
E       AssertionError: Attributes of DataFrame.iloc[:, 0] (column name="valid_long") are different
E       
E       Attribute "dtype" are different
E       [left]:  Int64
E       [right]: int64

tests/system/test_gbq.py:197: AssertionError</pre></details>

### LINKED ISSUE #431: tests.system.test_gbq.TestReadGBQIntegration: test_should_properly_handle_null_boolean failed
This test failed!

To configure my behavior, see [the Flaky Bot documentation](https://github.com/googleapis/repo-automation-bots/tree/main/packages/flakybot).

If I'm commenting on this issue too often, add the `flakybot: quiet` label and
I will stop commenting.

---

commit: 928e47b4bc1a8d933f48206ba98e3d22726876c8
buildURL: [Build Status](https://source.cloud.google.com/results/invocations/bbab8f02-1708-462f-8c90-4285e1304f0a), [Sponge](http://sponge2/bbab8f02-1708-462f-8c90-4285e1304f0a)
status: failed
<details><summary>Test output</summary><br><pre>self = <system.test_gbq.TestReadGBQIntegration object at 0x7f3e624f2ee0>
project_id = 'precise-truck-742'

    def test_should_properly_handle_null_boolean(self, project_id):
        query = "SELECT BOOLEAN(NULL) AS null_boolean"
        df = gbq.read_gbq(
            query,
            project_id=project_id,
            credentials=self.credentials,
            dialect="legacy",
        )
>       tm.assert_frame_equal(df, DataFrame({"null_boolean": [None]}))
E       AssertionError: Attributes of DataFrame.iloc[:, 0] (column name="null_boolean") are different
E       
E       Attribute "dtype" are different
E       [left]:  boolean
E       [right]: object

tests/system/test_gbq.py:436: AssertionError</pre></details>

### LINKED ISSUE #432: tests.system.test_gbq.TestReadGBQIntegration: test_should_properly_handle_nullable_booleans failed
This test failed!

To configure my behavior, see [the Flaky Bot documentation](https://github.com/googleapis/repo-automation-bots/tree/main/packages/flakybot).

If I'm commenting on this issue too often, add the `flakybot: quiet` label and
I will stop commenting.

---

commit: 928e47b4bc1a8d933f48206ba98e3d22726876c8
buildURL: [Build Status](https://source.cloud.google.com/results/invocations/bbab8f02-1708-462f-8c90-4285e1304f0a), [Sponge](http://sponge2/bbab8f02-1708-462f-8c90-4285e1304f0a)
status: failed
<details><summary>Test output</summary><br><pre>self = <system.test_gbq.TestReadGBQIntegration object at 0x7f3e62480310>
project_id = 'precise-truck-742'

    def test_should_properly_handle_nullable_booleans(self, project_id):
        query = """SELECT * FROM
                    (SELECT BOOLEAN(TRUE) AS nullable_boolean),
                    (SELECT NULL AS nullable_boolean)"""
        df = gbq.read_gbq(
            query,
            project_id=project_id,
            credentials=self.credentials,
            dialect="legacy",
        )
>       tm.assert_frame_equal(
            df, DataFrame({"nullable_boolean": [True, None]}).astype(object)
        )
E       AssertionError: Attributes of DataFrame.iloc[:, 0] (column name="nullable_boolean") are different
E       
E       Attribute "dtype" are different
E       [left]:  boolean
E       [right]: object

tests/system/test_gbq.py:448: AssertionError</pre></details>

### LINKED ISSUE #433: tests.system.test_gbq.TestReadGBQIntegration: test_one_row_one_column failed
This test failed!

To configure my behavior, see [the Flaky Bot documentation](https://github.com/googleapis/repo-automation-bots/tree/main/packages/flakybot).

If I'm commenting on this issue too often, add the `flakybot: quiet` label and
I will stop commenting.

---

commit: 928e47b4bc1a8d933f48206ba98e3d22726876c8
buildURL: [Build Status](https://source.cloud.google.com/results/invocations/bbab8f02-1708-462f-8c90-4285e1304f0a), [Sponge](http://sponge2/bbab8f02-1708-462f-8c90-4285e1304f0a)
status: failed
<details><summary>Test output</summary><br><pre>self = <system.test_gbq.TestReadGBQIntegration object at 0x7f3e627e8670>
project_id = 'precise-truck-742'

    def test_one_row_one_column(self, project_id):
        df = gbq.read_gbq(
            "SELECT 3 as v",
            project_id=project_id,
            credentials=self.credentials,
            dialect="standard",
        )
        expected_result = DataFrame(dict(v=[3]))
>       tm.assert_frame_equal(df, expected_result)
E       AssertionError: Attributes of DataFrame.iloc[:, 0] (column name="v") are different
E       
E       Attribute "dtype" are different
E       [left]:  Int64
E       [right]: int64

tests/system/test_gbq.py:633: AssertionError</pre></details>

### LINKED ISSUE #434: tests.system.test_gbq.TestReadGBQIntegration: test_query_with_parameters failed
This test failed!

To configure my behavior, see [the Flaky Bot documentation](https://github.com/googleapis/repo-automation-bots/tree/main/packages/flakybot).

If I'm commenting on this issue too often, add the `flakybot: quiet` label and
I will stop commenting.

---

commit: 928e47b4bc1a8d933f48206ba98e3d22726876c8
buildURL: [Build Status](https://source.cloud.google.com/results/invocations/bbab8f02-1708-462f-8c90-4285e1304f0a), [Sponge](http://sponge2/bbab8f02-1708-462f-8c90-4285e1304f0a)
status: failed
<details><summary>Test output</summary><br><pre>self = <system.test_gbq.TestReadGBQIntegration object at 0x7f3e626aa1f0>
project_id = 'precise-truck-742'

    def test_query_with_parameters(self, project_id):
        sql_statement = "SELECT @param1 + @param2 AS valid_result"
        config = {
            "query": {
                "useLegacySql": False,
                "parameterMode": "named",
                "queryParameters": [
                    {
                        "name": "param1",
                        "parameterType": {"type": "INTEGER"},
                        "parameterValue": {"value": 1},
                    },
                    {
                        "name": "param2",
                        "parameterType": {"type": "INTEGER"},
                        "parameterValue": {"value": 2},
                    },
                ],
            }
        }
        # Test that a query that relies on parameters fails
        # when parameters are not supplied via configuration
        with pytest.raises(ValueError):
            gbq.read_gbq(
                sql_statement,
                project_id=project_id,
                credentials=self.credentials,
                dialect="legacy",
            )
    
        # Test that the query is successful because we have supplied
        # the correct query parameters via the 'config' option
        df = gbq.read_gbq(
            sql_statement,
            project_id=project_id,
            credentials=self.credentials,
            configuration=config,
            dialect="legacy",
        )
>       tm.assert_frame_equal(df, DataFrame({"valid_result": [3]}))
E       AssertionError: Attributes of DataFrame.iloc[:, 0] (column name="valid_result") are different
E       
E       Attribute "dtype" are different
E       [left]:  Int64
E       [right]: int64

tests/system/test_gbq.py:722: AssertionError</pre></details>

### LINKED ISSUE #435: tests.system.test_gbq.TestReadGBQIntegration: test_struct failed
This test failed!

To configure my behavior, see [the Flaky Bot documentation](https://github.com/googleapis/repo-automation-bots/tree/main/packages/flakybot).

If I'm commenting on this issue too often, add the `flakybot: quiet` label and
I will stop commenting.

---

commit: 928e47b4bc1a8d933f48206ba98e3d22726876c8
buildURL: [Build Status](https://source.cloud.google.com/results/invocations/bbab8f02-1708-462f-8c90-4285e1304f0a), [Sponge](http://sponge2/bbab8f02-1708-462f-8c90-4285e1304f0a)
status: failed
<details><summary>Test output</summary><br><pre>self = <system.test_gbq.TestReadGBQIntegration object at 0x7f3e623f9e20>
project_id = 'precise-truck-742'

    def test_struct(self, project_id):
        query = """SELECT 1 int_field,
                   STRUCT("a" as letter, 1 as num) struct_field"""
        df = gbq.read_gbq(
            query,
            project_id=project_id,
            credentials=self.credentials,
            dialect="standard",
        )
        expected = DataFrame(
            [[1, {"letter": "a", "num": 1}]], columns=["int_field", "struct_field"],
        )
>       tm.assert_frame_equal(df, expected)
E       AssertionError: Attributes of DataFrame.iloc[:, 0] (column name="int_field") are different
E       
E       Attribute "dtype" are different
E       [left]:  Int64
E       [right]: int64

tests/system/test_gbq.py:847: AssertionError</pre></details>

### LINKED ISSUE #436: tests.system.test_gbq.TestReadGBQIntegration: test_array_length_zero failed
This test failed!

To configure my behavior, see [the Flaky Bot documentation](https://github.com/googleapis/repo-automation-bots/tree/main/packages/flakybot).

If I'm commenting on this issue too often, add the `flakybot: quiet` label and
I will stop commenting.

---

commit: 928e47b4bc1a8d933f48206ba98e3d22726876c8
buildURL: [Build Status](https://source.cloud.google.com/results/invocations/bbab8f02-1708-462f-8c90-4285e1304f0a), [Sponge](http://sponge2/bbab8f02-1708-462f-8c90-4285e1304f0a)
status: failed
<details><summary>Test output</summary><br><pre>self = <system.test_gbq.TestReadGBQIntegration object at 0x7f3e626b66d0>
project_id = 'precise-truck-742'

    def test_array_length_zero(self, project_id):
        query = """WITH t as (
                   SELECT "a" letter, [""] as array_field
                   UNION ALL
                   SELECT "b" letter, [] as array_field)
    
                   select letter, array_field, array_length(array_field) len
                   from t
                   order by letter ASC"""
        df = gbq.read_gbq(
            query,
            project_id=project_id,
            credentials=self.credentials,
            dialect="standard",
        )
        expected = DataFrame(
            [["a", [""], 1], ["b", [], 0]], columns=["letter", "array_field", "len"],
        )
>       tm.assert_frame_equal(df, expected)
E       AssertionError: Attributes of DataFrame.iloc[:, 2] (column name="len") are different
E       
E       Attribute "dtype" are different
E       [left]:  Int64
E       [right]: int64

tests/system/test_gbq.py:879: AssertionError</pre></details>

### LINKED ISSUE #437: tests.system.test_gbq.TestReadGBQIntegration: test_array_of_floats failed
This test failed!

To configure my behavior, see [the Flaky Bot documentation](https://github.com/googleapis/repo-automation-bots/tree/main/packages/flakybot).

If I'm commenting on this issue too often, add the `flakybot: quiet` label and
I will stop commenting.

---

commit: 928e47b4bc1a8d933f48206ba98e3d22726876c8
buildURL: [Build Status](https://source.cloud.google.com/results/invocations/bbab8f02-1708-462f-8c90-4285e1304f0a), [Sponge](http://sponge2/bbab8f02-1708-462f-8c90-4285e1304f0a)
status: failed
<details><summary>Test output</summary><br><pre>self = <system.test_gbq.TestReadGBQIntegration object at 0x7f3e62238700>
project_id = 'precise-truck-742'

    def test_array_of_floats(self, project_id):
        query = """select [1.1, 2.2, 3.3] as a, 4 as b"""
        df = gbq.read_gbq(
            query,
            project_id=project_id,
            credentials=self.credentials,
            dialect="standard",
        )
>       tm.assert_frame_equal(df, DataFrame([[[1.1, 2.2, 3.3], 4]], columns=["a", "b"]))
E       AssertionError: Attributes of DataFrame.iloc[:, 1] (column name="b") are different
E       
E       Attribute "dtype" are different
E       [left]:  Int64
E       [right]: int64

tests/system/test_gbq.py:911: AssertionError</pre></details>

### PR DESCRIPTION
**Pull request by tswast**
deps: ensure compatibility with google-cloud-bigquery 3.x

Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [ ] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/python-bigquery-pandas/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes #426
Fixes #437
Fixes #436
Fixes #435
Fixes #434
Fixes #433
Fixes #432
Fixes #431
Fixes #430
Fixes #429
🦕

### CODE REVIEW

**Comment by tswast** [comment]
Prerelease tests pass now! https://source.cloud.google.com/results/invocations/c11210cb-d8b9-4d4d-a07a-dfa03fbf8060/targets

File: tests/system/test_gbq.py [hunk 1]
```diff
@@ -10,7 +10,7 @@
 import numpy as np
 import pandas
 import pandas.api.types
-import pandas.util.testing as tm
+import pandas.testing as tm
```
Comment (plamut) [inline tests/system/test_gbq.py:13]: What does "tm" stand for?

File: pandas_gbq/gbq.py [hunk 2]
```diff
@@ -585,6 +585,7 @@ def _bqschema_to_nullsafe_dtypes(schema_fields):
         "DATETIME": "datetime64[ns]",
         "FLOAT": np.dtype(float),
         "GEOMETRY": "object",
+        "INTEGER": "Int64",
```
Comment (plamut) [inline pandas_gbq/gbq.py:588]: As per comment above, it seems we need to update `docs/source/reading.rst` as well?
(for INTEGER and BOOLEAN)

**Review by plamut — COMMENTED** [review]
Docs might need an update (according to a comment), but if not, the comment should be updated instead.

File: tests/system/test_gbq.py (hunk 1 above)
Comment (tswast) [reply tests/system/test_gbq.py:13]: Maybe "test module"? It's one of the leftovers from the pandas repo. e.g.

https://github.com/pandas-dev/pandas/blob/059c8bac51e47d6eaaa3e36d6a293a22312925e6/pandas/tests/arrays/categorical/test_sorting.py#L8

Comment (plamut) [reply tests/system/test_gbq.py:13]: Ah, it's test module, now I remember.

(I'm generally not a fan of possibly-cryptic abbreviations :slightly_smiling_face: )

**Review by plamut — APPROVED** [review]

**Pull request merged**